### PR TITLE
ViewDataAttribute

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewExecutor.cs [hunk 1]
```diff
@@ -257,5 +259,17 @@ public class ViewExecutor
                 await writer.FlushAsync();
             }
         }
+
+        protected virtual void OnExecuting(ViewContext viewContext)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewExecutor.cs]: What do you think about using the same pattern we have with `ITempDataCallback`? The name needs work, but the pattern is pretty straight forward.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataAttributePropertyProvider.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataAttributePropertyProvider.cs]: can this be real-internal?

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ViewDataAttributePropertyProvider.cs]: We want to use this in the Pages.

**Comment by pranavkm** [comment]
🆙 📅 

Missing unit tests, I'll add once we're ok with the way this is set up.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj [hunk 3]
```diff
@@ -18,20 +18,10 @@
 
   <ItemGroup>
     <!-- For testing the testing infrastructure only -->
-    <WebApplicationFactoryContentRootAttribute
-      Include="BasicWebSite"
-      AssemblyName="BasicWebsite"
-      ContentRootPath="../../../../WebSites/BasicWebSite"
-      ContentRootTest="BasicWebSite.csproj"
-      Priority="-1" />
+    <WebApplicationFactoryContentRootAttribute Include="BasicWebSite" AssemblyName="BasicWebsite" ContentRootPath="../../../../WebSites/BasicWebSite" ContentRootTest="BasicWebSite.csproj" Priority="-1" />
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj]: VS being VS. Let's not fight it

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/Infrastructure/HttpClientExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+using AngleSharp.Dom.Html;
+using AngleSharp.Parser.Html;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Mvc.FunctionalTests
+{
+    public static class HttpClientExtensions
+    {
+        public static async Task<IHtmlDocument> GetHtmlDocumentAsync(this HttpClient client, string requestUri)
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/Infrastructure/HttpClientExtensions.cs]: cc @javiercn

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/IPageExecutionCallbackFeature.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages
+{
+    internal interface IPageExecutionCallbackFeature
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/IPageExecutionCallbackFeature.cs]: What do you like better about this? I thought we were adding a feature specifically for view data?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageViewDataAttributeFilter.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        public void OnPageHandlerExecuted(PageHandlerExecutedContext context)
+        {
+        }
+
+        public void OnPageHandlerExecuting(PageHandlerExecutingContext context)
+        {
+            Subject = context.HandlerInstance;
+            context.HttpContext.Features.Set<IPageExecutionCallbackFeature>(this);
+            context.HttpContext.Features.Set<IViewComponentExecutionCallbackFeature>(this);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageViewDataAttributeFilter.cs]: This doesn't compose by the way. You can only have one of these if it's used like this.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/IPageExecutionCallbackFeature.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/IPageExecutionCallbackFeature.cs]: Ahh, I thought all we settled on was making this a feature.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageViewDataAttributeFilter.cs (hunk 6 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageViewDataAttributeFilter.cs]: They're meant to be internal contracts for now and we don't have a reason to compose them as yet. Do you foresee a reason to make them composable?

**Comment by pranavkm** [comment]
@rynowak could you have a look? I made a small change to the feature contract + added tests.

**Comment by javiercn** [comment]
LGTM, get Ryan’s sign off

**Pull request merged**